src/packages/engine-core/src/NAPIEngine.ts

```diff
@@ -0,0 +1,536 @@
+        this.config.clientVersion!,
+      )
+    }
+    return platform
+  }
+
+  private convertDatasources(
+    datasources: DatasourceOverwrite[],
+  ): Record<string, string> {
+    const obj = Object.create(null)
+    for (const { name, url } of datasources) {
+      obj[name] = url
+    }
+    return obj
+  }
+  private async loadEngine(): Promise<void> {
+    if (!this.engine) {
+      if (!this.QueryEngine) {
+        if (!this.libQueryEnginePath) {
+          this.libQueryEnginePath = await this.getLibQueryEnginePath()
+          debug(`using ${this.libQueryEnginePath}`)
+        }
+        try {
+          this.QueryEngine = require(this.libQueryEnginePath).QueryEngine
+        } catch (e) {
+          // How should we handle this
+          throw new Error(e)
+        }
+      }
+      if (this.QueryEngine) {
+        try {
+          this.engine = new this.QueryEngine(
+            {
+              datamodel: this.datamodel,
+              datasourceOverrides: this.datasourceOverrides,
+              logLevel: this.config.logLevel ?? 'off',
+            },
+            (err, log) => {
+              if (err) throw new Error(err)
+              debug(`received next log event`)
+              let event: QueryEngineEvent | null = null
+              try {
+                event = JSON.parse(log)
+                if (!event) return
+              } catch (e) {
+                throw new Error(e)
+              }
+              event.level = event?.level.toLowerCase() ?? 'unknown'
+              if (isQueryEvent(event)) {
+                this.logEmitter.emit('query', {
+                  timestamp: Date.now(),
+                  query: event.query,
+                  params: event.params,
+                  duration: event.duration_ms,
+                  target: event.module_path,
+                })
+              } else {
+                this.logEmitter.emit(event.level, event)
+              }
+            },
+          )
+        } catch (e) {
+          const error = this.parseInitError(e.message)
+          if (typeof error === 'string') {
+            throw e
+          } else {
+            throw new PrismaClientInitializationError(
+              error.message,
+              this.config.clientVersion!,
+              error.error_code,
+            )
+          }
+        }
+      }
+    }
+  }
+
+  private parseInitError(str: string): SyncRustError | string {
+    try {
+      const error = JSON.parse(str)
+      if (typeof error.is_panic !== 'undefined') {
+        return error
+      }
+    } catch (e) {
+      //
+    }
+    return str
+  }
+  private parseRequestError(str: string): RustRequestError | string {
+    try {
+      const error = JSON.parse(str)
+      if (typeof error.is_panic !== 'undefined') {
+        return error
+      }
+    } catch (e) {
+      //
+    }
+    return str
+  }
+
+  on(event: EngineEventType, listener: (args?: any) => any): void {
+    if (event === 'beforeExit') {
```

Comment:
Is this still relevant

Reply:
I've implemented this on disconnect and kill

Reply:
Should still be relevant for users https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#hooks
```
Runs before the Prisma process exits and allows you to perform additional queries - for example, as part of a graceful shutdown.
```

Reply:
Implemented